bgpd: fix invalid ipv4-vpn nexthop for IPv6 peer

proelbtn opened the pull request:
Given that two routers are connected to each other, they have IPv6 addresses,  and they establish BGP peer with extended-nexthop capability and one router tries to advertise locally-generated IPv4-VPN routes to another router, bgpd on the router that tries to advertise IPv4-VPN routes will be crashed with "invalid MP nexthop length (AFI IP6)".

```
2022/11/12 09:17:31 BGP: [J85T9-K7RBY][EC 33554504] bpacket_reformat_for_peer: fd00:1:1:2::1: invalid MP nexthop length (AFI IP6): 0
BGP: Received signal 11 at 1668244651 (si_addr 0x0, PC 0x7f4c0c1e8334); aborting...
BGP: /lib/libfrr.so.0(zlog_backtrace_sigsafe+0x71) [0x7f4c0c1b1ce1]
BGP: /lib/libfrr.so.0(zlog_signal+0xf9) [0x7f4c0c1b1ee9]
BGP: /lib/libfrr.so.0(+0xd6875) [0x7f4c0c1de875]
BGP: /lib/x86_64-linux-gnu/libpthread.so.0(+0x14420) [0x7f4c0bf5d420]
BGP: /lib/libfrr.so.0(stream_fifo_push+0x14) [0x7f4c0c1e8334]
BGP: /usr/lib/frr/bgpd(+0x13c6ca) [0x5606a34206ca]
BGP: /usr/lib/frr/bgpd(bgp_generate_updgrp_packets+0xf0) [0x5606a34229b0]
BGP: /lib/libfrr.so.0(thread_call+0x81) [0x7f4c0c1f0ae1]
BGP: /lib/libfrr.so.0(frr_run+0xe8) [0x7f4c0c1aa138]
BGP: /usr/lib/frr/bgpd(main+0x385) [0x5606a33ca9f5]
BGP: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f4c0bd7b083]
BGP: /usr/lib/frr/bgpd(_start+0x2e) [0x5606a33cceae]
BGP: in thread (bgp_generate_updgrp_packets) scheduled from bgpd/bgp_fsm.c:996 bgp_adjust_routeadv()
```

This issue happened because MP_REACH_NLRI path attribute is not generated correctly when ipv4-vpn routes are advertised to IPv6 peer. When IPv4 routes are leaked from VRF RIB, the nexthop of these routes are also IPv4 address (0.0.0.0/0 or specific addresses). However, bgp_packet_mpattr_start only covers the case of IPv6 nexthop (for IPv6 peer) and bpacket_reformat_for_peer reads 0 (for SNPA) to check nexthop length, resulting bgpd crashes.

ipv4-unicast routes were not affected by this issue because the case of IPv4 nexthop is covered in `else` block. So, this PR changes the sequence of `if` blocks to the same for UNICAST routes.

Signed-off-by: Ryoga Saito <ryoga.saito@linecorp.com>

ton31337 reviewed (commented):
Can you cover this in topotests?

ton31337 commented:
@mergifyio backport stable/8.4

ton31337 reviewed (approved):
LGTM, please just apply @frrbot styling.

riw777 reviewed (approved):
looks good

proelbtn commented:
@ton31337 @riw777 I fixed the style error and I added a new topotest that covers this issue. You can reproduce the same issue with this topotest and unmodified FRR.

```
cat /tmp/topotests/bgp_srv6l3vpn_over_ipv6.test_bgp_srv6l3vpn_over_ipv6/r1/bgpd.*
2022/11/17 11:29:50 BGP: [T83RR-8SM5G] bgpd 8.5-dev-verda-gUNKNOWN starting: vty@2605, bgp@<all>:179
2022/11/17 11:29:50 BGP: [GQF43-N30BN] bgp_install_info_to_zebra: No zebra instance to talk to, not installing information
2022/11/17 11:29:50 BGP: [GQF43-N30BN] bgp_install_info_to_zebra: No zebra instance to talk to, not installing information
2022/11/17 11:29:50 BGP: [YAF85-253AP][EC 100663299] buffer_flush_available: write error on fd 2: Bad file descriptor
2022/11/17 11:29:50 BGP: [YAF85-253AP][EC 100663299] buffer_flush_available: write error on fd 2: Bad file descriptor
2022/11/17 11:29:50 BGP: [M7Q4P-46WDR] vty[28]@> enable
2022/11/17 11:29:50 BGP: [M7Q4P-46WDR] vty[28]@# clear log cmdline-targets
2022/11/17 11:29:50 BGP: [M7Q4P-46WDR] vty[28]@(config)# log commands
2022/11/17 11:29:50 BGP: [M7Q4P-46WDR] vty[28]@(config)# log timestamp precision 3
2022/11/17 11:29:52.613 BGP: [YXTCF-6ENSA][EC 33554510] Interface: eth0 does not have a v6 LL address associated with it, waiting until one is created for it
2022/11/17 11:29:52.613 BGP: [YXTCF-6ENSA][EC 33554510] Interface: eth0 does not have a v6 LL address associated with it, waiting until one is created for it
2022/11/17 11:29:52.613 BGP: [MVZKX-EG443][EC 33554452] bgp_process_packet: BGP OPEN receipt failed for peer: 2001::2
2022/11/17 11:29:52.614 BGP: [YXTCF-6ENSA][EC 33554510] Interface: eth0 does not have a v6 LL address associated with it, waiting until one is created for it
2022/11/17 11:29:52.614 BGP: [YXTCF-6ENSA][EC 33554510] Interface: eth0 does not have a v6 LL address associated with it, waiting until one is created for it
2022/11/17 11:29:53.715 BGP: [JR3R6-FJ4P2][EC 33554505] bpacket_reformat_for_peer: 2001::2: invalid MP nexthop length (AFI IP6): 0
BGP: Received signal 11 at 1668652193 (si_addr 0x0, PC 0x7fb7f58d1984); aborting...
BGP: zlog_signal+0xdf                   7fb7f589f31f     7ffc21a7dbf0 /usr/lib/frr/libfrr.so.0 (mapped at 0x7fb7f5800000)
BGP: core_handler+0xa3                  7fb7f58c8da3     7ffc21a7dd30 /usr/lib/frr/libfrr.so.0 (mapped at 0x7fb7f5800000)
BGP: _L_unlock_13+0x34                  7fb7f340f630     7ffc21a7de80 /lib64/libpthread.so.0 (mapped at 0x7fb7f3400000)
BGP:     ---- signal ----
BGP: stream_fifo_push+0x14              7fb7f58d1984     7ffc21a7eb80 /usr/lib/frr/libfrr.so.0 (mapped at 0x7fb7f5800000)
BGP: bgp_packet_add+0x5b                      4772cb     7ffc21a7eb90 /usr/lib/frr/bgpd (mapped at 0x400000)
BGP: bgp_generate_updgrp_packets+0xc0           4790b0     7ffc21a7ebd0 /usr/lib/frr/bgpd (mapped at 0x400000)
BGP: thread_call+0x89                   7fb7f58d9649     7ffc21a7ec60 /usr/lib/frr/libfrr.so.0 (mapped at 0x7fb7f5800000)
BGP: frr_run+0xd0                       7fb7f5897610     7ffc21a7ecf0 /usr/lib/frr/libfrr.so.0 (mapped at 0x7fb7f5800000)
BGP: main+0x35c                               426bcc     7ffc21a7ee70 /usr/lib/frr/bgpd (mapped at 0x400000)
BGP: __libc_start_main+0xf5             7fb7f3022555     7ffc21a7eed0 /lib64/libc.so.6 (mapped at 0x7fb7f3000000)
BGP: _start+0x29                              4291ab     7ffc21a7ef90 /usr/lib/frr/bgpd (mapped at 0x400000)
BGP: in thread bgp_generate_updgrp_packets scheduled from bgpd/bgp_fsm.c:580 bgp_routeadv_timer()
```

ton31337 commented on tests/topotests/bgp_srv6l3vpn_over_ipv6/r1/bgpd.conf:
```diff
@@ -0,0 +1,52 @@
+log commands
+!
+router bgp 65001
+ bgp router-id 1.1.1.1
```
Can you replace this with http://docs.frrouting.org/projects/dev-guide/en/latest/topotests.html (^f+`RFC 5737`)

proelbtn commented:
@ton31337 Sorry, I fixed.

proelbtn commented:
ci:rerun

riw777 reviewed (approved):
looks good
